pkgs/development/compilers/swift/compiler/default.nix

```diff
@@ -0,0 +1,555 @@
+{ lib
+, stdenv
+, cmake
+, coreutils
+, gnugrep
+, perl
+, ninja
+, pkg-config
+, clang
+, bintools
+, python3
+, git
+, fetchFromGitHub
+, makeWrapper
+, gnumake
+, file
+, runCommand
+, writeShellScriptBin
+# Linux-specific
+, glibc
+, libuuid
+# Darwin-specific
+, substituteAll
+, fixDarwinDylibNames
+, xcbuild
+, sigtool
+, DarwinTools
+, CoreServices
+, Foundation
+, Combine
+, CLTools_Executables
+}:
+
+let
+
+  inherit (stdenv) hostPlatform targetPlatform;
+
+  # The Swift toolchain script builds projects with separate repos. By convention, some of them share
+  # the same version with the main Swift compiler project per release.
+  version = "5.6.2";
+
+  fetchSwiftRelease = { repo, sha256  }:
+    fetchFromGitHub {
+      owner = "apple";
+      inherit repo sha256;
+      rev = "swift-${version}-RELEASE";
+      name = "${repo}-${version}-src";
+    };
+
+  # Names in this set match the directory the source is unpacked to.
+  sources = {
+    cmark = fetchSwiftRelease {
+      repo = "swift-cmark";
+      sha256 = "0340j9x2n40yx61ma2pgqfbn3a9ijrh20iwzd1zxqq87rr76hh3z";
+    };
+    llvm-project = fetchSwiftRelease {
+      repo = "llvm-project";
+      sha256 = "131i1n5zmy0k0k56cpp6baiajg5zprip894npsl5ciknlna3fnv1";
+    };
+    swift = fetchSwiftRelease {
+      repo = "swift";
+      sha256 = "1418y8z27c4gx8alhmhjp64gcfmpggh46476rdpc54l4fl0mf962";
+    };
+  }
+    // lib.optionalAttrs (!stdenv.isDarwin) {
+      swift-corelibs-libdispatch = fetchSwiftRelease {
+        repo = "swift-corelibs-libdispatch";
+        sha256 = "1vv4y2rw6c5ixqndy6sqc17dc5rhsh1hv1cnpmkvjz978f8jrlnn";
+      };
+    };
+
+  # There are apparently multiple naming conventions on Darwin. Swift uses the
+  # xcrun naming convention. See `configure_sdk_darwin` calls in CMake files.
+  swiftOs = if targetPlatform.isDarwin
+    then {
+      "macos" = "macosx";
+      "ios" = "iphoneos";
+      #iphonesimulator
+      #appletvos
+      #appletvsimulator
+      #watchos
+      #watchsimulator
+    }.${targetPlatform.darwinPlatform}
```

Comment:
Not 100% necessary buy a default via `or` would be nice to have.

Reply:
Hmm, I'm not sure that'd even build. Isn't an error at evaluation time better in this case?

It looks like we do have some iOS target support, but I didn't actually test it. Other than that, this set is already broader than what we support in Nixpkgs.

Reply:
`abort` with a helpful message also counts as a default in my books ;)

An attr not found exception cannot be caught in any way and isn't very user-friendly.

Reply:
I added:

```
or (throw "Cannot build Swift for target Darwin platform '${targetPlatform.darwinPlatform}'")
```